### PR TITLE
feat(select): align selected option over trigger

### PR DESCRIPTION
**Pull request by kara**
r: @jelbourn

### CODE REVIEW

File: src/lib/select/select.ts [hunk 1]
```diff
@@ -22,7 +22,35 @@ import {Subscription} from 'rxjs/Subscription';
 import {transformPlaceholder, transformPanel, fadeInContent} from './select-animations';
 import {ControlValueAccessor, NgControl} from '@angular/forms';
 import {coerceBooleanProperty} from '../core/coersion/boolean-property';
-import {ConnectedOverlayPositionChange} from '../core/overlay/position/connected-position';
+import {ConnectedOverlayDirective} from '../core/overlay/overlay-directives';
+import {ViewportRuler} from '../core/overlay/position/viewport-ruler';
+
```
Comment (jelbourn) [inline src/lib/select/select.ts]: Maybe add a high level comment here explaining that you need to know a handful of style values here in order to properly align the selected item on open.

File: src/lib/select/select.ts [hunk 2]
```diff
@@ -375,4 +425,169 @@ export class MdSelect implements AfterContentInit, ControlValueAccessor, OnDestr
     }, undefined);
   }
 
+  /** Calculates the scroll position and x- and y-offsets of the overlay panel. */
+  private _calculateOverlayPosition(): void {
+    this._offsetX = this._isRtl() ? SELECT_PANEL_PADDING_X : -SELECT_PANEL_PADDING_X;
+
+    const panelHeight = this.options.length <= SELECT_MAX_OPTIONS_DISPLAYED
```
Comment (jelbourn) [inline src/lib/select/select.ts]:
```ts
Math.min(this.options.length * SELECT_OPTION_HEIGHT,  SELECT_PANEL_MAX_HEIGHT)
```
?

File: src/lib/select/select.ts [hunk 3]
```diff
@@ -375,4 +425,169 @@ export class MdSelect implements AfterContentInit, ControlValueAccessor, OnDestr
...
+    return optionOffsetFromPanelTop * -1 - heightDiff;
+  }
+
+  /**
+   * Checks that the attempted overlay position will fit within the viewport.
+   * If it will not fit, tries to adjust the scroll position and the associated
+   * y-offset so the panel can open fully on-screen. If it still won't fit,
+   * sets the offset back to 0 to allow the fallback position to take over.
+   */
+  private _checkOverlayWithinViewport(maxScroll: number): void {
```
Comment (jelbourn) [inline src/lib/select/select.ts]: Why isn't the behavior
"default position" > "fallback position" ( * n) > slide to stay on-screen

With the last one being part of the connected position strategy?

File: src/lib/select/select.ts [hunk 4]
```diff
@@ -375,4 +425,169 @@ export class MdSelect implements AfterContentInit, ControlValueAccessor, OnDestr
...
+      this._scrollTop = maxScroll;
+      this._offsetY = 0;
+      this._transformOrigin = `50% top 0px`;
+      return;
+    }
+  }
+
+  /** Sets the transform origin point based on the selected option. */
+  private _getOriginBasedOnOption(): string {
+    return `50% ${Math.abs(this._offsetY) - 9 + SELECT_OPTION_HEIGHT / 2}px 0px`;
```
Comment (jelbourn) [inline src/lib/select/select.ts]: `- 9`?

Comment (kara) [inline src/lib/select/select.ts]: Whoops, that shouldn't be hardcoded. It's the heightDiff number. I'll fix!

File: src/lib/select/select.ts (hunk 3 above)
Comment (kara) [inline src/lib/select/select.ts]: (discussed over IM)

File: src/lib/select/select.spec.ts [hunk 5]
```diff
@@ -490,79 +490,276 @@ describe('MdSelect', () => {
...
+    let select: HTMLElement;
+
+    beforeEach(() => {
+      fixture = TestBed.createComponent(BasicSelect);
+      fixture.detectChanges();
+      trigger = fixture.debugElement.query(By.css('.md-select-trigger')).nativeElement;
+      select = fixture.debugElement.query(By.css('md-select')).nativeElement;
+    });
+
+    function checkTriggerAlignedWithOption(index: number): void {
```
Comment (jelbourn) [inline src/lib/select/select.spec.ts]: Add comment
```ts
  /**
   * Asserts that the given option is aligned with the trigger.
   * @param index The index of the option.
   */
```

**Comment by kara** [comment]
@jelbourn Comments addressed

**Comment by jelbourn** [comment]
LGTM

**Pull request merged**